core/src/outstanding_requests.rs

```diff
@@ -0,0 +1,271 @@
+use crate::request_response::RequestResponse;
```

Comment:
nit: `outstanding_requests` feels too generic to me in core: `outstanding_repair_requests`.  If we had that `repair` crate then calling this module `outstanding_requests` would be fine 😵

Reply:
@mvines, so the way it's currently set up, `outstanding_requests::OutstandingRequests<T, S>` is designed to be more general than just repair, it can track any request type `T` that implements `T: RequestResponse<Response = S>`.

For instance gossip could potentially initialize an `OutstandingRequests<T, S>` for gossip requests/responses as well.

Reply:
Is that a thing we need though?   Building generic interfaces in the hopes that somebody in the future will use them when there's only a single consumer today usually doesn't end too well.

Reply:
@mvines yeah I would agree in mosts cases, but there was a lot of talk of a similar solution to solve the gossip replay issue here: https://github.com/solana-labs/solana/issues/9491. 

It was a also a nice way to logically bundle the verification of responses into the actual request object like this: https://github.com/solana-labs/solana/pull/9903/files#diff-dc108ad584300fa57f99d1675ccff3b9R66, so that they fit together rather than in disparate functions.